src/bluesearch/database/topic_info.py

```diff
@@ -0,0 +1,116 @@
+# Blue Brain Search is a text mining toolbox focused on scientific use cases.
+#
+# Copyright (C) 2020  Blue Brain Project, EPFL.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Lesser General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public License
+# along with this program. If not, see <https://www.gnu.org/licenses/>.
+"""Implementation of the TopicInfo data structure."""
+from __future__ import annotations
+
+import copy
+import datetime
+import pathlib
+from dataclasses import dataclass, field
+from typing import Any
+
+import bluesearch
+from bluesearch.database.article import ArticleSource
+
+
+@dataclass
+class TopicInfo:
+    """The topic information extracted from a journal article.
+
+    For the spec see the following GitHub issue/comment:
+    https://github.com/BlueBrain/Search/issues/518#issuecomment-985525160
+    """
+
+    source: ArticleSource
+    path: str | pathlib.Path
+    element_in_file: int | None = None
+    article_topics: dict[str, list[str]] = field(init=False, default_factory=dict)
+    journal_topics: dict[str, list[str]] = field(init=False, default_factory=dict)
+    metadata: dict = field(default_factory=dict)
```

Comment:
Might be nice to document these in the class docstring under `Attributes` (I guess that is the right section).

Reply:
Just tried and learned that `sphinx`/`autodoc`/`napoleon` don't play well with dataclasses. They're not clever enough to recognise them and document everything as class variables. If one includes a class docstring with `Parameters` and `Attributes` then sphinx generates duplicate entries and even issues warnings for `Attributes` because they are already implicitly included in the doc.

I haven't found a way to override this behaviour, there's a very ugly way of inline-documenting attributes/class variables:
```python
    source: ArticleSource
    path: str | pathlib.Path
    element_in_file: int | None = None
    #: A mapping of the form topic-kind -> list-of-topics that contains
    #: article topics. For example: ``{"MeSH": ["topic 1", "topic 2"]}``.
    #: Use ``add_article_topics`` method to populate it.
    article_topics: dict[str, list[str]] = field(init=False, default_factory=dict)
    #: A mapping of the form topic-kind -> list-of-topics that contains
    #: journal topics. For example: ``{"MeSH": ["topic 1", "topic 2"]}``.
    #: Use ``add_journal_topics`` method to populate it.
    journal_topics: dict[str, list[str]] = field(init=False, default_factory=dict)
```
but to my eyes this looks extremely ugly.

I can think of two ways of going forward:
* Not documenting attributes of dataclasses at all
* Documenting paramters in the class docstring (but they will be duplicated as class variables because that's what sphinx thinks they are) and attributes inline like in the example above.

Any suggestions?

Reply:
Pretty weird. I did not know about this issue. Yeh, maybe we don't have to write any docstrings. As you want:) Feel free to resolve without any action.

Reply:
I'd leave it like it is for now. I think it relatively transparent right now. Maybe if we add more stuff to this class (cf. the metadata discussion) then it might be worth adding more documentation about it too.